packages/wallet-comm/src/wallet-functions.ts

```diff
@@ -0,0 +1,163 @@
+import BN from 'bn.js';
+import { WalletEventListener } from '.';
+
+
+export function createWalletIFrame(iframeId: string, walletUrl: string): HTMLIFrameElement {
+  const iFrame = document.createElement("iframe");
+  iFrame.src = walletUrl;
+  iFrame.id = iframeId;
+  iFrame.style.display = 'none';
+  iFrame.style.position = 'absolute';
+  iFrame.style.left = '0px';
+  iFrame.style.right = '0px';
+  iFrame.style.bottom = '0px';
+  iFrame.style.top = '0px';
+  iFrame.width = '0';
+  iFrame.height = '0';
+  iFrame.style.zIndex = '9999';
+  iFrame
+
+  window.addEventListener('message', (event => {
+    if (event.data && event.data.type && event.data.type === SHOW_WALLET) {
+      iFrame.style.display = 'initial';
+      document.body.style.overflow = 'hidden';
+      iFrame.width = '100%';
+      iFrame.height = '100%';
+    }
+    if (event.data && event.data.type && event.data.type === HIDE_WALLET) {
+      iFrame.style.display = 'none';
+      document.body.style.overflow = 'initial';
+      iFrame.width = '0';
+      iFrame.height = '0';
+
+    }
+  }));
+  return iFrame;
+}
+
+
+/**
+ * Initialized the wallet with a given user id.
+ * The promise resolves when the wallet is initialized or an error occurs
+ * The promise returns the wallet address.
+ */
+export async function initializeWallet(iFrameId: string, userId: string): Promise<string> {
+  const iFrame = <HTMLIFrameElement>document.getElementById(iFrameId);
+  const message = initializeRequest(userId);
+  iFrame.contentWindow.postMessage(message, "*");
+  return new Promise((resolve, reject) => {
+    window.addEventListener("message", function eventListener(event: MessageEvent) {
+      if (event.data && event.data.type && (
+        event.data.type === INITIALIZATION_SUCCESS || event.data.type === INITIALIZATION_FAILURE)) {
+
+        window.removeEventListener("message", eventListener);
+        if (event.data.type === INITIALIZATION_SUCCESS) {
+          resolve(event.data.address);
+        } else {
+          reject(event.data.message);
+        }
+      }
+    });
+  });
+}
+
+
+/**
+ * Opens the channel.
+ */
+//TODO: Can this be part of funding instead of it's own method
+export function openChannel(iFrameId: string, channel: Channel): void {
+  const iFrame = <HTMLIFrameElement>document.getElementById(iFrameId);
+  const message = openChannelRequest(channel);
+  iFrame.contentWindow.postMessage(message, "*");
+
+}
+/**
+ *  Validates signed data.
+ * Promise resolves when the data is verified or an error occurs.
+ */
+export async function validateSignature(iFrameId: string, data, signature: string): Promise<boolean> {
+  const iFrame = <HTMLIFrameElement>document.getElementById(iFrameId);
+  const message = validationRequest(data, signature);
+  iFrame.contentWindow.postMessage(message, "*");
```

Comment:
Can there exist a race condition where the response event is fired before the listener is added?

Reply:
That's a good point, we should probably create the promise before posting the message